File: python/day1.py
from utilities import read_int_input


def two_sum(data, target):
    s = set()
    for number in data:
        other = target - number
        if other in s:
            return other * number
        s.add(number)
    return None


def part_1(data):
    return two_sum(data, 2020)


def part_2(data):
    for index, number in enumerate(data):
        result = two_sum(data[index+1:], 2020 - number)
        if result:
            return number * result


def fix_expense_report():
    print('Day 1')
    data = read_int_input('../inputs/day1.txt')
    print('\tPart 1: {}'.format(part_1(data)))
    print('\tPart 2: {}'.format(part_2(data)))


if __name__ == '__main__':
    fix_expense_report()

File: rust/src/day1.rs
use crate::utilities;
use std::collections::HashSet;

pub fn two_sum(data: &[i32], target: i32) -> Option<i32> {
    let mut set = HashSet::new();
    for &number in data {
        let other = target - number;
        match set.get(&other) {
            None => {
                set.insert(number);
            }
            Some(_) => {
                return Some(number * other);
            }
        }
    }
    None
}

fn part_1(data: &[i32]) {
    println!("\tPart 1: {}", two_sum(data, 2020).unwrap());
}

fn part_2(data: &[i32]) {
    let mut result = 0;
    for (index, &number) in data.iter().enumerate() {
        if let Some(other) = two_sum(&data[index + 1..], 2020 - number) {
            result = other * number;
            break;
        }
    }
    if result != 0 {
        println!("\tPart 2: {}", result);
    }
}

pub fn main() {
    println!("Day 1");
    let data = utilities::read_int_array("../inputs/day1.txt");
    part_1(&data);
    part_2(&data);
}
